testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/javascript/JavascriptAdapterTest.java

```diff
@@ -400,6 +398,58 @@ public void loginRequiredAction() {
                 .init(defaultArguments(), this::assertSuccessfullyLoggedIn);
     }
 
+    /**
+     * Test for scope handling via {@code initOptions}: <pre>{@code
+     * Keycloak keycloak = new Keycloak(); keycloak.init({.... scope: "profile email phone"})
+     * }</pre>
+     * See KEYCLOAK-14412
+     */
+    @Test
+    public void testScopeInInitOptionsShouldBeConsideredByLoginUrl() {
+
+        JSObjectBuilder initOptions = defaultArguments()
+                .loginRequiredOnLoad()
+                // phone is optional client scope
+                .add("scope", "profile email phone");
+
+        try {
+            testExecutor.init(initOptions);
+            // This throws exception because when JavascriptExecutor waits for AsyncScript to finish
+            // it is redirected to login page and executor gets no response
+
+            throw new RuntimeException("Probably the login-required OnLoad mode doesn't work, because testExecutor should fail with error that page was redirected.");
+        } catch (WebDriverException ex) {
+            // should happen
+        }
+
+        testExecutor.loginForm(testUser, this::assertOnTestAppUrl)
+                .init(initOptions, this::assertSuccessfullyLoggedIn)
+        .executeScript("return window.keycloak.tokenParsed.scope", assertOutputContains("phone"));
+    }
+
+    /**
+     * Test for scope handling via {@code loginOptions}: <pre>{@code
+     * Keycloak keycloak = new Keycloak(); keycloak.login({.... scope: "profile email phone"})
+     * }</pre>
+     * See KEYCLOAK-14412
+     */
+    @Test
+    public void testScopeInLoginOptionsShouldBeConsideredByLoginUrl() {
+
+        try {
+            testExecutor.configure()
+                        .init(defaultArguments());
+        } catch (WebDriverException ex) {
+            // should happen
+        }
+
+        JSObjectBuilder loginOptions = JSObjectBuilder.create().add("scope", "profile email phone");
+
+        testExecutor.login(loginOptions.build(), (JavascriptStateValidator) (driver, output, events) -> {
+            assertThat(jsDriver.getCurrentUrl(), containsString("&scope=openid%20profile%20email%20phone"));
```

Comment:
The JavascriptAdapter tests are a little bit confusing in this. They are mixing two variable names `driver` and `jsDriver`. Out testsuite has two browsers because JS adapter doesn't work in Htmlunit, which is the default browser for majority of tests (it is default because it is fast). So we added `jsDriver` as the second browser, which defaults to Phantomjs. This browser is able to work with JS adapter. 

Now the `JavascriptStateValidator` is there for checking the state after some Javascript piece of code was executed. It has three parameters `(driver, output, events)`. At the moment, the `driver` parameter is maybe a little bit redundant here as we always use `jsDriver`, but this can change in the future. So it would be better to use it here like this:
```suggestion
            assertThat(driver.getCurrentUrl(), containsString("&scope=openid%20profile%20email%20phone"));
```

this way you use a webdriver provided by testExecutor and function (in this case `login()`) can decide what browser to use for validating the condition.


One more thing, you don't need to use the `testExecutor` for everything, the `jsDriver` (do not use `driver` outside of) is available even outside of testExecutor and it is perfectly ok to use it.

Reply:
Hopefully, it is understandable and I don't just confuse you more. I like our JS tests, but lots of people are struggling with them. It is almost done though,  just some nitpicks to fix